### ISSUE TITLE
Support for the uws module

### ISSUE THREAD

**Issue by henopied** [Barbosik/MultiOgar#158]
I would like to see `uws` support as it could allow many more clients to connect to multiogar while using less system resources.

**Comment by MastaCoder** [comment]
"`µWS` is one of the most lightweight, efficient & scalable WebSocket server implementations available." - I'm interested :+1:

**Comment by NatsuTheGreat** [comment]
I've tried. Had problems with a compiler or something. Gave up.

**Comment by henopied** [comment]
The issue is with the sendpacket function.

**Comment by NatsuTheGreat** [comment]
I don't see what can be causing it. :/

**Comment by Influxx** [comment]
I also tried implementing uws but couldn't get past the sendPacket issue. Maybe @Barbosik can help out or give us a clue?

**Comment by David1ali12** [comment]
`uws` so much faster

**Comment by henopied** [comment]
I have found the issue and that is the uws client is not exposed so we cannot add things to its prototype. https://github.com/uWebSockets/uWebSockets/issues/184

**Comment by xinyuweb** [comment]
del    "WebSocket.prototype.sendPacket"

add ws.sendPacket to function "close"


	function close(error) {
			this.server.ips[this.socket.remoteAddress] -= 1;
			// Log disconnections
			this.server.log.onDisconnect(this.socket.remoteAddress);

			var client = this.socket.playerTracker;
			var len = this.socket.playerTracker.cells.length;
			for (var i = 0; i < len; i++) {
				var cell = this.socket.playerTracker.cells[i];

				if (!cell) {
					continue;
				}

				cell.calcMove = function () {
					return;
				}; // Clear function so that the cell cant move
				this.server.removeNode(cell);
			}
			client.disconnect = this.server.config.playerDisconnectTime * 20;
			this.socket.sendPacket = function () {
				return;
			}; // Clear function so no packets are sent

			if (client.disconnect == -20) {
				var index = this.server.clients.indexOf(this.socket)
					if (index > -1) {
						this.server.clients.splice(index, 1);
					}
			}
		}

		ws.sendPacket = function (packet) {
			function getBuf(data) {
				var array = new Uint8Array(data.buffer || data);
				var l = data.byteLength || data.length;
				var o = data.byteOffset || 0;
				var buffer = new Buffer(l);

				for (var i = 0; i < l; i++) {
					buffer[i] = array[o + i];
				}

				return buffer;
			}

			//if (this.readyState == WebSocket.OPEN && (this._socket.bufferSize == 0) && packet.build) {
			if (this.readyState == 1 && packet.build) {
				var buf = packet.build();
				this.send(getBuf(buf), {
					binary : true
				});
			} else if (!packet.build) {
				// Do nothing
			} else {
				this.readyState = WebSocket.CLOSED;
				this.emit('close');
				this.removeAllListeners();
			}
		}

		ws.remoteAddress = ws._socket.remoteAddress;
		ws.remotePort = ws._socket.remotePort;
		this.log.onConnect(ws.remoteAddress); // Log connections

		ws.playerTracker = new PlayerTracker(this, ws);
		ws.packetHandler = new PacketHandler(this, ws);
		ws.on('message', ws.packetHandler.handleMessage.bind(ws.packetHandler));
		var bindObject = {
			server : this,
			socket : ws
		};

		ws.on('error', close.bind(bindObject));
		ws.on('close', close.bind(bindObject));
		this.clients.push(ws);
	}

**Comment by NatsuTheGreat** [comment]
^^ That can get use uws?

**Comment by henopied** [comment]
Please don't do that...

**Comment by henopied** [comment]
The above code wouldn't really be ideal.

**Comment by alexhultman** [comment]
Could you please try out 0.7.2? I think it should be fixed.

**Comment by Influxx** [comment]
Thanks @henopied and @alexhultman !

**Comment by NatsuTheGreat** [comment]
Is there a tutorial for windows 10?

**Comment by alexhultman** [comment]
@NatsuTheGreat of uws? Nope it doesn't work on Windows currently. There are patches pending to be merged to nodejs which will allow it to work though.

**Comment by NatsuTheGreat** [comment]
@alexhultman RIP. Going to have to stick with ```ws```.

**Comment by alexhultman** [comment]
@NatsuTheGreat That's up to you. Next Node.js 6.x version will have the patches though.

**Comment by NatsuTheGreat** [comment]
I am using v6.2.1

**Comment by alexhultman** [comment]
@NatsuTheGreat Right, I forgot. It actually already works on 6.3.0 but not with SSL. SSL will come with the next version after 6.3.0.

**Comment by NatsuTheGreat** [comment]
I am not using SSL, I have an older modified version of MultiOgar. Should I just get 6.3.0?

**Comment by henopied** [comment]
Should I make a pull request that allows one to choose between uws and ws in the config?

**Comment by David1ali12** [comment]
omg yes @henopied  please

**Comment by ItzLevvie** [comment]
@henopied You've to delete the Ogar repo that you've forked long time ago and then fork MultiOgar

**Comment by henopied** [comment]
@ItzLevvie Oh whoops, silly me!

**Comment by Barbosik** [comment]
does it has any advantage against ws?

**Comment by alexhultman** [comment]
@Barbosik https://github.com/uWebSockets/uWebSockets#benchmarks-table---validate

**Comment by NatsuTheGreat** [comment]
```
        throw new Error('Compilation of µWebSockets has failed and there is no pre-compiled binary ' +
        ^
 
Error: Compilation of µWebSockets has failed and there is no pre-compiled binary available for your system. Please install a supported C++11 compiler and reinstall the module 'uws'.
```

**Comment by alexhultman** [comment]
@NatsuTheGreat I assume you did not read the installation instructions?

**Comment by NatsuTheGreat** [comment]
Setting it up on windows is confusing me..

**Comment by SimonOrJ** [comment]
I thought Windows support was broken. You need knowledge of C++ or something to use on windows.

#158#issuecomment-232164992

**Comment by NatsuTheGreat** [comment]
It doesn't say that support for windows is broken anymore.

**Comment by alexhultman** [comment]
Yes it is *very* confusing to run:

npm install windows-build-tools
npm install uws

especially when this is clearly written in English instructions on the front page under Installation.

**Comment by NatsuTheGreat** [comment]
Lmao Alex, and what about those compilers you were talking about?

> On Windows you need to search the web for pre-compiled binaries or simply compile the dependencies yourself.

Very explanatory. ```npm install windows-build-tools``` Doesn't make a difference as of now.

**Comment by alexhultman** [comment]
Installation -> Node.js is your section, not the "native developers" section...

Never mind, clearly I will have to bundle it with pre-built binaries for Windows...

**Comment by NatsuTheGreat** [comment]
Well lets see, Alex:

>npm install --save uws

First of what I see there, and then:

>npm install windows-build-tools

I have installed

>npm install uws

And 

> npm install windows-build-tools

Yet the error keeps talking about a C++ Compiler. :/

**Comment by Barbosik** [comment]
@alexhultman: does it provide any functionality to minimize impact of DDoS attacks?

**Comment by alexhultman** [comment]
@NatsuTheGreat You did *reinstall* uws after installing windows-build-tools?

**Comment by NatsuTheGreat** [comment]
I went in order,

>npm install uws
npm install windows-build-tools

**Comment by alexhultman** [comment]
@NatsuTheGreat That statement does not answer my question, besides - it is the opposite of what the installation instructions tells you to. You need to *reinstall* uws after installing windows-build-tools.

**Comment by NatsuTheGreat** [comment]
Okay, sorry. I'll try it now.

**Comment by NatsuTheGreat** [comment]
Same error. There are also problems with ```npm install windows-build-tools```

**Comment by alexhultman** [comment]
You can write in my Gitter.

**Comment by NatsuTheGreat** [comment]
Oh, I had to run that through Admin CMD Prompt. Python was installed. Waiting for more replies from console.

**Comment by alexhultman** [comment]
Write in Gitter, I guess you write `npm install --global --production windows-build-tools`

**Comment by Megabyte918** [comment]
@alexhultman vape naysh yall :+1:
